crates/core/src/operations/put.rs

```diff
@@ -400,18 +400,41 @@ impl Operation for PutOp {
                 }
                 PutMsg::SuccessfulPut { id, .. } => {
                     match self.state {
-                        Some(PutState::AwaitingResponse { key, upstream }) => {
+                        Some(PutState::AwaitingResponse { key, upstream, contract, state }) => {
+                            tracing::debug!(tx = %id, %key, peer = %op_manager.ring.connection_manager.get_peer_key().unwrap(), "Contract not cached @ peer, caching after successful put");
+                            // Always seed the contract locally after a successful put
+                            op_manager.ring.seed_contract(key);
+
+                            tracing::debug!(tx = %id, %key, peer = %op_manager.ring.connection_manager.get_peer_key().unwrap(), "Storing contract locally");
+                            let contract_stored = match op_manager
+                                .notify_contract_handler(ContractHandlerEvent::PutQuery {
+                                    key,
+                                    state,
+                                    related_contracts: RelatedContracts::default(),
+                                    contract: Some(contract.clone()),
+                                })
+                                .await
+                            {
+                                Ok(_) => true,
+                                Err(err) => {
+                                    tracing::error!(tx = %id, %key, error = %err, "Failed to store contract locally");
+                                    false
+                                }
+                            };
+
                             let is_subscribed_contract = op_manager.ring.is_seeding_contract(&key);
-                            if !is_subscribed_contract && op_manager.ring.should_seed(&key) {
-                                tracing::debug!(tx = %id, %key, peer = %op_manager.ring.connection_manager.get_peer_key().unwrap(), "Contract not cached @ peer, caching");
+                            if !is_subscribed_contract {
+                                tracing::debug!(tx = %id, %key, peer = %op_manager.ring.connection_manager.get_peer_key().unwrap(), "Peer not subscribed to contract, starting subscription request");
+                                // This will make the node subscribe to the contract without doing a get since we're already seeding it locally
```

Comment:
the only thign is missing I think, in this case, to make the put atomic, is when we request a subscribe, is subordinating the completion of this operation to the completion of the subscribe (and we should probably do the same for puts when they require a preliminary get)

but I don't think we can link one transaction to an other transaction result right now, and it will require further changes.

If is the case, let's leave a TODO here for that change and tvkle it later